docs/adr/ADR-003-async-only-api-with-blocking-get.md

```diff
@@ -0,0 +1,449 @@
+# ADR-003: Async-Only API with Blocking GET Pattern
+
+**Status:** Accepted
+**Date:** 2025-11-19
+**Authors:** cassing
+**Supersedes:** ADR-003 v1 (direct-vs-queued-execution-modes)
+**Superseded by:** N/A
+
+## Context
+
+The qontract-api serves two distinct use cases with different requirements:
+
+### Use Case 1: MR Validation (GitLab Merge Requests)
+
+- **Purpose**: Validate proposed changes before merging
+- **Requirements**:
+  - **Quick response**: MR job needs result within minutes (< 5 minutes)
+  - **Dry-run only**: Never apply changes, only calculate actions
+  - **Result retrieval**: Job must get the calculated actions
+  - **Log output**: Display actions in MR comment
+
+### Use Case 2: Production Runs (Scheduled Jobs)
+
+- **Purpose**: Apply changes to external systems
+- **Requirements**:
+  - **Background execution**: Can take longer (> 10 minutes okay)
+  - **Apply changes**: Actually modify external systems
+  - **Fire-and-forget**: Client doesn't need immediate result
+  - **Status tracking**: Optionally check progress via task_id
+
+### Problem
+
+Initial approach (ADR-003 v1) used an `execution_mode` parameter:
+
+- `execution_mode="direct"`: Execute synchronously, return result
+- `execution_mode="queued"`: Queue task, return task_id
+
+**Issues with parameter-based approach:**
+
+1. **Two execution paths**: Router had branching logic for direct vs queued
+2. **Inconsistent responses**: Different response types based on parameter
+3. **FastAPI timeout risk**: Direct mode could timeout for long operations
+4. **Unclear semantics**: Parameter-based control is less RESTful
+
+**Better approach**: Use HTTP verbs to control behavior
+
+- **POST**: Create task (always async, always returns task_id)
+- **GET**: Retrieve result (blocks until complete or timeout)
+
+## Decision
+
+**Use async-only API pattern where POST creates tasks and GET retrieves results with blocking behavior.**
+
+### POST: Always Async (Queue Task)
+
+All reconciliation POST requests queue a background background task and return immediately.
+
+```http
+POST /api/v1/integrations/slack-usergroups/reconcile
+Content-Type: application/json
+
+{
+  "workspaces": [...],
+  "dry_run": true
+}
+
+Response: 202 Accepted
+{
+  "task_id": "550e8400-e29b-41d4-a716-446655440000",
+  "status": "queued",
+  "status_url": "/api/v1/integrations/slack-usergroups/reconcile/550e8400-e29b-41d4-a716-446655440000"
+}
```

Comment:
Should we include `Location` in Response header too?
* https://www.rfc-editor.org/rfc/rfc7240#section-4.1
* https://learn.microsoft.com/en-us/azure/architecture/best-practices/api-design#implement-asynchronous-methods

Reply:
Good idea. Added in [650c86c](https://github.com/app-sre/qontract-reconcile/pull/5320/commits/650c86c5994ca64f0750820769d777db5b01cd70)